ArduCopter/mode_auto.cpp

```diff
@@ -408,6 +414,64 @@ void ModeAuto::circle_start()
     }
 }
 
+// returns true when we have completed doing loiter turns
+// false if we still have remaining tasks
+// also takes us to next states after we complete the current state task 
+bool Mode::_Circling::is_completed()
+{
+    switch (_state) {
+    case _Circling::State::MOVING_TO_EDGE:
+        if (copter.wp_nav->reached_wp_destination()) {
+            circling_start();
+        }
+        break;
+    case _Circling::State::CIRCLING:
+        // check if we have completed circling
+        if (fabsf(copter.circle_nav->get_angle_total()/M_2PI) >= _total_turns) {
+            // stop at circle edge where we started
+            copter.wp_nav->set_wp_destination_loc(circle_edge);
+            return true;
+        }
+        break;
+    }
+
+    return false;
+}
```

Comment:
```suggestion
    switch (_state) {
    case _Circling::State::MOVING_TO_EDGE:
        if (!copter.wp_nav->reached_wp_destination()) {
            return false;
        }
        circling_start();  // changes _state to _Circling::State::Circling
        FALLTHROUGH;
    case _Circling::State::CIRCLING:
        // check if we have completed circling
        if (fabsf(copter.circle_nav->get_angle_total()/M_2PI) < _total_turns) {
            return false;
        }
        // stop at circle edge where we started
        copter.wp_nav->set_wp_destination_loc(circle_edge);
        FALLTHROUGH;
    }

    return true;
}
```

... might be nice to have a `_Circling::State::FINISHED` so we don't repeatedly set the waypoint destination after finishing the circle.

Reply:
Done! I have updated this method to  also do what we were earlier doing in `is_completed()`. Have also introduced this new state named `FINISHED`.